### PR TITLE
feat: add operator user

### PR DESCRIPTION
**Pull request by DeannaSharma**
- Created operator user group
- Added custom attribute "nocCodes" to user pool
- created new function "createUserWithCustomAttribute" to handle the creation of an operator user with NOC codes
- Updated add user page to include operator user group, when operator user is selected at least one NOC code must be provided
- Updated user management pages for org admin and sys admin to correctly render operator user types in user table
- added properties "isOperatorUser" and "nocCodes" to idtoken so that it can be brought back in session detail for future tickets
- refactored tests to use "mockSession" and "mockSessionWithOrgDetail" where necessary

### CODE REVIEW

File: site/schemas/add-user.schema.ts [hunk 1]
```diff
@@ -2,14 +2,39 @@ import { UserGroups } from "@create-disruptions-data/shared-ts/enums";
...
 export const addUserSchema = z.object({
     givenName: z.string(setZodDefaultError("Enter a first name")).min(1),
     familyName: z.string(setZodDefaultError("Enter a last name")).min(1),
     email: z.string(setZodDefaultError("Enter a valid email address")).email(),
     orgId: z.string().uuid(),
     group: z.nativeEnum(UserGroups, setZodDefaultError("Select which account is required")),
+    operatorNocCodes: z.array(operatorDataSchema).optional(),
 });
 
+export const addUserSchemaRefined = z
```
[the remark this answers is not in the document]
Comment (DeannaSharma) [reply site/schemas/add-user.schema.ts:22]: done

File: site/data/cognito.ts [hunk 2]
```diff
@@ -279,6 +279,58 @@ export const deleteUsersByAttribute = async (attributeName: string, attributeVal
     }
 };
 
+export const createUserWithCustomAttribute = async (
```
[the remark this answers is not in the document]
Comment (DeannaSharma) [reply site/data/cognito.ts:282]: done

File: site/pages/api/admin/add-user.api.ts [hunk 3]
```diff
@@ -12,6 +13,24 @@ import {
 } from "../../../utils/apiUtils";
 import { getSession } from "../../../utils/apiUtils/auth";
 
+export const formatAddUserBody = (body: object) => {
+    const operatorNocCodes = Object.entries(body)
+        .filter((item) => item.toString().startsWith("operatorNocCodes"))
```
Comment (ljones92) [inline site/pages/api/admin/add-user.api.ts:18]:
```suggestion
        .filter((item) => item[0].startsWith("operatorNocCodes"))
```

File: site/pages/api/admin/add-user.api.ts [hunk 4]
```diff
@@ -12,6 +13,24 @@ import {
...
+        .map((arr: string[]) => {
+            const [, values] = arr;
+            return JSON.parse(values) as OperatorData;
+        });
+
+    const cleansedBody = Object.fromEntries(
+        Object.entries(body).filter((item) => !item.toString().startsWith("operatorNocCodes")),
```
Comment (ljones92) [inline site/pages/api/admin/add-user.api.ts:25]:
```suggestion
        Object.entries(body).filter((item) => !item[0].startsWith("operatorNocCodes")),
```

File: site/pages/api/admin/add-user.api.ts [hunk 5]
```diff
@@ -35,11 +56,24 @@ const addUser = async (req: NextApiRequest, res: NextApiResponse) => {
             return;
         }
 
-        await createUser(validatedBody.data);
+        if (validatedBody.data.group === UserGroups.operators) {
+            const nocCodes = validatedBody.data.operatorNocCodes?.map((operator) => operator.nocCode) ?? [];
+            const operatorAttribute: AttributeType[] = [
+                {
+                    Name: "custom:nocCodes",
+                    Value: nocCodes.toString(),
```
Comment (ljones92) [inline site/pages/api/admin/add-user.api.ts:64]: `nocCodes.join(",")` is probably the more idiomatic way of concatenating the strings here

File: site/schemas/add-user.schema.ts [hunk 6]
```diff
@@ -2,14 +2,37 @@ import { UserGroups } from "@create-disruptions-data/shared-ts/enums";
...
+export const addUserSchemaRefined = addUserSchema
+    .refine(
+        (input) => {
+            return !(input.group === UserGroups.operators && input.operatorNocCodes?.length === 0);
+        },
+        { path: ["operatorNocCodes"], message: "Select at least one NOC" },
+    )
+    .refine(
+        (input) => {
+            return !(input.operatorNocCodes !== undefined && input.operatorNocCodes.length > 5);
```
[the remark this answers is not in the document]
Comment (DeannaSharma) [reply site/schemas/add-user.schema.ts:31]: done

File: site/pages/admin/add-user.page.tsx [hunk 7]
```diff
@@ -2,30 +2,113 @@ import { UserGroups } from "@create-disruptions-data/shared-ts/enums";
...
+import { flattenZodErrors, sortOperatorByName } from "../../utils";
 import { destroyCookieOnResponseObject, getPageState } from "../../utils/apiUtils";
 import { getSessionWithOrgDetail } from "../../utils/apiUtils/auth";
 import { getStateUpdater } from "../../utils/formUtils";
 
 const title = "Add User - Create Transport Disruptions Service";
 const description = "Add User page for the Create Transport Disruptions Service";
 
-export interface AddUserPageProps extends PageState<Partial<AddUserSchema>> {}
+const filterConfig = {
```
[the remark this answers is not in the document]
Comment (DeannaSharma) [reply site/pages/admin/add-user.page.tsx:28]: removed filterOptions entirely from SearchSelect component. removing it doesn't seem to affect the search input

File: site/pages/admin/add-user.page.tsx [hunk 8]
```diff
@@ -2,30 +2,113 @@ import { UserGroups } from "@create-disruptions-data/shared-ts/enums";
...
+    const getOperatorRows = () => {
+        if (pageState.inputs.operatorNocCodes) {
+            return sortOperatorByName(pageState.inputs.operatorNocCodes).map((operator) => ({
+                cells: [
+                    `${operator.nocCode} - ${operator.operatorPublicName}`,
+                    <button
+                        id={`remove-service-${operator.nocCode}`}
+                        key={`remove-service-${operator.nocCode}`}
+                        className="govuk-link"
+                        onClick={(e) => Promise.resolve(removeOperator(e, operator.nocCode))}
```
[the remark this answers is not in the document]
Comment (DeannaSharma) [reply site/pages/admin/add-user.page.tsx:102]: removed

File: site/pages/api/admin/add-user.api.ts (hunk 3 above)
Comment (DeannaSharma) [reply site/pages/api/admin/add-user.api.ts:18]: done

File: site/pages/api/admin/add-user.api.ts (hunk 4 above)
Comment (DeannaSharma) [reply site/pages/api/admin/add-user.api.ts:25]: done

File: site/pages/api/admin/add-user.api.ts (hunk 5 above)
Comment (DeannaSharma) [reply site/pages/api/admin/add-user.api.ts:64]: done

**Review by ljones92 — APPROVED** [review]

**Pull request merged**